### PR TITLE
widlcard_matching readability and Space Complexity

### PR DESCRIPTION
**Pull request by sparsh44**
### Describe your change:
**Increased Readability** of 1 if  else condition (line no. 77) and added the **Space Complexity** of the algorithm (line no. 16). 

* [ ] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: dynamic_programming/wildcard_matching.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
+"""
+Implementation of regular expression matching with support for '.' and '*'.
+'.' Matches any single character.
+'*' Matches zero or more of the preceding element.
+The matching should cover the entire input string (not partial).
+
```
Comment (CaedenPH) [inline dynamic_programming/wildcard_matching.py:6]:
```suggestion
```

File: dynamic_programming/wildcard_matching.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+"""
+
+
+def match_pattern(input_string: str, pattern: str) -> bool:
+    """
+    uses bottom-up dynamic programming solution for matching the input
```
Comment (CaedenPH) [inline dynamic_programming/wildcard_matching.py:12]:
```suggestion
    Uses bottom-up dynamic programming solution for matching the input
```

**Comment by sparsh44** [comment]
@CaedenPH I was thinking if this code can be further optimized in terms of space complexity or not!!

File: dynamic_programming/wildcard_matching.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    # given pattern.
+    # "dp" stands for dynamic programming.
+    dp = [[0 for i in range(len_pattern)] for j in range(len_string)]
+
+    # since string of zero length match pattern of zero length
+    dp[0][0] = 1
+
+    # since pattern of zero length will never match with string of non-zero length
+    for i in range(1, len_string):
+        dp[i][0] = 0
```
Comment (realstealthninja) [inline dynamic_programming/wildcard_matching.py:71]: Wonder if a map could be used to increase performance?

File: dynamic_programming/wildcard_matching.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+                if dp[i][j - 2] == 1:
+                    dp[i][j] = 1
+                elif pattern[j - 2] in (input_string[i - 1], "."):
+                    dp[i][j] = dp[i - 1][j]
+                else:
+                    dp[i][j] = 0
+            else:
+                dp[i][j] = 0
+
+    return bool(dp[-1][-1])
```
Comment (realstealthninja) [inline dynamic_programming/wildcard_matching.py:97]: Why are the indexes in the negative?

File: dynamic_programming/wildcard_matching.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+            elif pattern[j - 1] == "*":
+                if dp[i][j - 2] == 1:
+                    dp[i][j] = 1
+                elif pattern[j - 2] in (input_string[i - 1], "."):
+                    dp[i][j] = dp[i - 1][j]
+                else:
+                    dp[i][j] = 0
+            else:
+                dp[i][j] = 0
```
Comment (realstealthninja) [inline dynamic_programming/wildcard_matching.py:95]: Way too many if and elif statement could probably convert a massive chunk of them in to guard clauses

File: dynamic_programming/wildcard_matching.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    for i in range(1, len_string):
+        dp[i][0] = 0
+
+    # since string of zero length will match with pattern where there
+    # is at least one * alternatively
+    for j in range(1, len_pattern):
+        if pattern[j - 1] == "*":
+            dp[0][j] = dp[0][j - 2]
+        else:
+            dp[0][j] = 0
```
Comment (realstealthninja) [inline dynamic_programming/wildcard_matching.py:79]:
```suggestion
        if pattern[j - 1] != "*":
            dp[0][j] = 0
            continue
        dp[0][j - 2]
```

File: dynamic_programming/wildcard_matching.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    # since string of zero length match pattern of zero length
+    dp[0][0] = 1
+
+    # since pattern of zero length will never match with string of non-zero length
+    for i in range(1, len_string):
+        dp[i][0] = 0
+
+    # since string of zero length will match with pattern where there
+    # is at least one * alternatively
+    for j in range(1, len_pattern):
```
Comment (realstealthninja) [inline dynamic_programming/wildcard_matching.py:75]: Cant you just check for * and . using the in keyword in python?

Comment (realstealthninja) [reply dynamic_programming/wildcard_matching.py:75]: Example:
```
if "*" in somestring:
    # i remember there being a built in func to find the index of a substring should look into it if you are going this route 
```

**Review by realstealthninja — CHANGES_REQUESTED** [review]

File: dynamic_programming/wildcard_matching.py (hunk 6 above)
Comment (sparsh44) [reply dynamic_programming/wildcard_matching.py:79]: I think the first one is easy to read and understand!!

File: dynamic_programming/wildcard_matching.py (hunk 7 above)
Comment (realstealthninja) [reply dynamic_programming/wildcard_matching.py:75]: Ah

**Pull request closed without merging**